### ISSUE TITLE
Stop supporting 32-bit

### ISSUE THREAD

**Issue by vittorioromeo** [SFML/SFML#1930]
Almost no one is using 32-bit architectures anymore (e.g. see [Steam hardware surveys](https://store.steampowered.com/hwsurvey/Steam-Hardware-Software-Survey-Welcome-to-Steam)), but we are still supporting it in our CI. We should drop support for obsolete architectures.

**Comment by Bromeon** [comment]
Before making such a call, we should probably gather some more data. The Steam page is definitely very helpful, but SFML has also been used for a handful of non-game applications.

As this might prevent some legit use cases on 32-bit platforms, we should also know more precisely what cost maintaining 32-bit systems has for us. We had now [the MinGW issue](https://github.com/SFML/SFML/pull/1922#issuecomment-1000427164), but were there many other instances? An alternative might also be to drop MinGW support, its toolchain has always been a second-class citizen in Windows, and several parts (such as linker) are quite bad to begin with.

TLDR: if there's an easy way to keep supporting 32-bit systems, we might want to consider it.

**Comment by eXpl0it3r** [comment]
I'm against removing architecture, platform or compiler support when there's no real reason to do so.

Yes, there's low usage, but we don't gain anything at this point in time, from nuking 32-bit support on Windows and the MinGW issue is unlikely due to a 32-bit architecture.

> An alternative might also be to drop MinGW support (at least for 32-bit), its toolchain has always been a second-class citizen in Windows, and several parts (such as linker) are quite bad to begin with.

Again, I wouldn't just drop a compiler, especially one that is still widely used. I feel like it's too often just dismissed, especially by those who think, the only way to program on Windows is with Visual Studio. 😄

**Comment by vittorioromeo** [comment]
> Before making such a call, we should probably gather some more data. 

I agree with that. I will look into more statistics, and ask around in the SFML Discord/forum to see if anyone is actively using SFML on a 32-bit platform.

> An alternative might also be to drop MinGW support

MSYS2 + MinGW is my main development environment on Windows, and I know many other developers who run a similar system. I personally find the MSVC and Visual Studio ecosystem quite dreadful.

> I'm against removing architecture, platform or compiler support when there's no real reason to do so.

The reason is to make SFML more lightweight and make the developers' life easier. I see it as "cleanup", pretty much similarly to why we are migrating to C++17 and dropping support for older compilers.

**Comment by vittorioromeo** [comment]
Oh, and most importantly -- this change would be 3.x only. People requiring to target 32-bit platforms would still have access to SFML 2.x!

**Comment by vittorioromeo** [comment]
More statistics increasing my confidence about 32-bit being obsolete in 2021:
https://www.pcbenchmarks.net/os-marketshare.html

**Comment by eXpl0it3r** [comment]
What these statistics all miss is that you can still run 32-bits applications on 64-bits OS. Check your task manager on Windows and start counting the number of `(32 bit)` tags on the process names.

It's not just a clean up and it doesn't make SFML itself more lightweight, all it does right now is remove a bunch of binaries and making it harder for those who want to build a 32-bit application. It right now costs us nothing to keep the support, but anyone who still wants to build SFML as 32-bit application, has to recompile all dependencies and add back some CMake code.

It's been discussed a long time ago, that the dependency management needs to move away from in-source binary shipping, to on-the-fly building. With that in mind as well, the "clean up" here only addresses some parts, while removing existing options that then have to be restored.

I honestly don't want to be the one who has to explain to people that we removed 32-bit support, because we determined nobody should use it anymore.

**Comment by vittorioromeo** [comment]
> What these statistics all miss is that you can still run 32-bits applications on 64-bits OS.

That's true, but I don't see how that's relevant. If you have a 64-bit OS, you can run both 32-bit and 64-bit applications. So if everybody uses a 64-bit architecture, there's no point in creating new software targeting 32-bit architecture. The support is just there for backwards compatibility.

> making it harder for those who want to build a 32-bit application

I'm not entirely certain that's a bad thing. And again, the same argument applies to using a more modern version of C++. You're effectively alienating a subset of users that have to use machines or architectures where C++17 is not available. And I'd reckon there are more of those compared to 32-bit architectures...

> It right now costs us nothing to keep the support

We have three jobs in CI to test 32-bit architectures. These take time/resources to execute.

> I honestly don't want to be the one who has to explain to people that we removed 32-bit support, because we determined nobody should use it anymore, as if we're the judge of how people should or shouldn't be building their applications.

We are already the judges of that in different places, most notably the minimum requirement of C++17 for SFML 3.x. There are always going to be people who are unhappy about the requirement or choices a library imposes, but we need to look forward and modernize our targets. SFML is already late to the C++ standard modernization party, do we want to spend a few more years actively supporting a very likely obsolete architecture like it has happened with C++03? 

I'd rather move forward with this proposal *unless* we get reports of people that are targeting 32-bit architectures using SFML and that would like to switch to SFML 3.x and keep targeting those architectures. My intuition tells me there's no one like that, but I am happy to be proven wrong. 

And, yet again, I want to reiterate that SFML 2.x will not go anywhere. If someone wants to live in the past (or needs to, unfortunately), SFML 2.x will still be there.

**Comment by Bromeon** [comment]
> I honestly don't want to be the one who has to explain to people that we removed 32-bit support, because we determined nobody should use it anymore, as if we're _the_ judge of how people should or shouldn't be building their applications.

What makes this even harder, is that we simply don't know. We can ask in the forum and on Discord, but many SFML users won't regularly check there -- _especially_ those who might be affected. Someone may have integrated SFML into a larger 32-bit application 2 years ago, and maybe update every now and then (non-breaking), and things have always just worked.

In a professional environment, it can be frustrating if things that have always worked suddenly stop working. But what's even more frustrating, is:
* The official reason given by the developers is "we wanted to clean up the library and came to the conclusion your setup was obsolete", i.e. neither a technical nor a personnel-driven reason why the setup is no longer supported.
* Unlike with API changes, there is no real way forward for the user. If people go for 32-bit, they likely have a reason to, and can't simply upgrade OS/processor architecture. "Stay with SFML 2" is not a solution because we will stop maintaining it.

---

> I'd rather move forward with this proposal unless we get reports of people that are targeting 32-bit architectures using SFML and that would like to switch to SFML 3.x and keep targeting those architectures. My intuition tells me there's no one like that, but I am happy to be proven wrong.

The problem with that is that you expect all users to regularly report their issues with SFML on GitHub. I know from my own experience that this is not the case. Many times, especially under time constraints, when evaluating libraries I would just check which one fulfills the requirements I have. If I'm already using a library and at some point it starts to make questionable design choices, I start looking for alternatives. This is particularly the case for people outside the typical open-source spectrum (i.e. commercial users), which don't have time to argue with developers about their philosophies.

---

> I'm not entirely certain that's a bad thing. And again, the same argument applies to using a more modern version of C++. You're effectively alienating a subset of users that have to use machines or architectures where C++17 is not available.

The reason why I consider C++17 a different case, is because migrating to a modern C++ standard benefits 95% of users, as it makes a) the library more robust (e.g. through consequent use of smart pointers) and b) the user experience better (by allowing them to use modern C++ in SFML's APIs).

We're not doing any favor to 64-bit users by abandoning 32-bit. The advantages we're gaining are our own (CI + not having to deal with rare 32-bit related questions).

**Comment by binary1248** [comment]
Is there a hard requirement somewhere that makes this necessary? I'm normally also not the kind of guy to drag old technology with me past their expiry date but completely throwing out entire architectures just based on (maybe not even representative) surveys isn't something I feel good about. As mentioned in the comments above, 32-bit applications are still supported on Windows and the last time I checked Raspberry Pi OS 64-bit is still considered beta.

If the idea is that we should encourage adoption of 64-bit IMO that is not our responsibility. That responsibility rests with the OS vendors/maintainers. We shouldn't assume that this decision is harder for us to make than it is for them. It is costing them every day they have to keep supporting 32-bit architectures so their decision to keep supporting them must have a pretty good reason.

Since C++ is in theory supposed to be an architecture independent language, I don't see what kind of a justification we would be able to provide users besides "we were just lazy". If "supporting" 32-bit is really costing us maintenance effort then maybe we should fix that problem instead. In a perfect world we shouldn't have to think about which architectures we target when we write code, picking data types based solely on the kind of data we intend to store.

Regarding CI: I don't think CI performance should ever be a reason for making changes to our code, especially not removing entire architectures. CI is really only there to provide us with information on whether the code still builds on selected platforms and that no regressions have sneaked in via a changeset once we have good test coverage. Alone the fact that the build time relies so much on available CPU/RAM makes the metric of total CI run time kind of moot. If we had unlimited resources we could rent many fat servers and make sure CI completes in under a minute, but what is the point? As soon as I see a CI failure on a specific platform I try my best to build locally using said compiler before pushing again so that I can react faster to any newer errors (can't do this for macOS since I don't have one). CI build time was never a bottleneck for me.

**Comment by vittorioromeo** [comment]
I understand we will not find consensus on this. Don't get me wrong -- you all have valid points. But I do believe you are overestimating how many users really need to target 32-bit. And those same few users, if existing, wouldn't likely care to update to SFML 3.x or use C++17 features. 

I'm quite confident nobody would care if we dropped 32-bit support from 3.x and only kept in 2.x, and no one has stepped forward yet with *"I would care because I target 32-bit"*. Regardless, this is not worth fighting over, so I will close the PR and issue. Maybe this will be revisited in a few years... when it will be too late :P 

I might still merge the PR in my C++17 branch for people who want to get rid of support for older architectures.
